content/blog/testthat-3-3-0/index.Rmd

```diff
@@ -0,0 +1,172 @@
+    #| error: true
+
+    test_that("show off expect_shape() failure messages", {
+      x <- matrix(1:9, nrow = 3)
+      expect_shape(x, nrow = 4)
+      expect_shape(x, dim = c(3, 3, 3))
+      expect_shape(x, dim = c(3, 4))
+    })
+    ```
+
+As you can see from the examples above, when you run a single test interactively (i.e. not as a part of a test suite) you now see exactly how many expectations succeeded and failed.
+
+## Other new features
+
+* testthat generally does a better job of handling nested tests, aka subtests, where you put a `test_that()` inside another `test_that()`, or more typically `it()` inside of `describe()`. Subtests will now generate more informative failure messages, free from duplication, with more informative skips if any subtests don't contain any expectations.
+
+* The snapshot experience has been significantly improved, with all known bugs fixed and some new helpers added: `snapshot_reject()` rejects all modified snapshots by deleting the `.new` variants, and `snapshot_download_gh()` makes it easy to get snapshots off GitHub and into your local package. Additionally, `expect_snapshot()` and friends will now fail when creating a new snapshot on CI, as that's usually a signal that you've forgotten to run the snapshot code locally before committing.
+
+* On CRAN, `test_that()` will automatically skip if a package is not installed, which means that you no longer need to check if suggested packages are installed in your tests.
+
+* `vignette("mocking")` explains mocking in detail, and new `local_mocked_s3_method()`, `local_mocked_s4_method()`, and `local_mocked_r6_class()` make it easier to mock S3 and S4 methods and R6 classes.
+
+* `test_dir()`, `test_check()`, and friends gain a `shuffle` argument that uses `sample()` to randomly reorder the top-level expressions in each test file. This random reordering surfaces dependencies between tests and code outside of any test, as well as dependencies between tests, helping you find and eliminate unintentional dependencies.
+
+* `try_again()` is now publicized, as it's a useful tool for testing flaky code:
+
+    ```{r}
+    #| eval: false
+
+    flaky_function <- function() {
+      if (runif(1) < 0.1) 0 else 1
+    }
+    test_that("my flaky test is ok", {
+      # 10% chance of failure:
+      expect_equal(flaky_function(), 1)
+
+      # 1% chance of failure:
+      try_again(1, expect_equal(flaky_function(), 1))
+
+      # 0.1% chance of failure:
+      try_again(2, expect_equal(flaky_function(), 1))
+    })
+    ```
+
+* New `skip_unless_r()` skips tests on unsuitable versions of R. It has a convenient syntax so you can use, e.g., `skip_unless_r(">= 4.1.0")` to skip tests that require `...names()`.
+
+* New `SlowReporter` makes it easier to find the slowest tests in your package. You can run it with `devtools::test(reporter = "slow")`.
+
+* New `vignette("challenging-functions")` provides an index to other documentation organized by various challenges.
+
```

Comment:
To use the new features, do you recommend bumping the version of testthat in Suggests? Might be a good place to mention it.

Unfortunately pkgload only checks for Imports not Suggests, so bumping the dep won't trigger the install prompt on load.

Reply:
Yeah, and `install.packages()` won't check for it either. So overall, I don't think it's worth it.